src/torchmetrics/functional/text/infolm.py

```diff
@@ -0,0 +1,653 @@
+        pad_token_id:
+            An id of ``<PAD>`` tokens that are used to make arrays of tokens the same size for batching purpose
+        cls_token_id:
+            An id of ``<CLS>`` token that represents the class of the input. (It might be ``<BOS>`` token for some
+            models.)
+        sep_token_id:
+            An id of ``<SEP>`` token that separates two different sentences in the same input. (It might be ``<EOS>``
+            token for some models.)
+
+    Return:
+        Tensor mask of 0s and 1s that masks all special tokens in the ``input_ids`` tensor.
+    """
+    token_mask = input_ids.eq(pad_token_id) | input_ids.eq(sep_token_id) | input_ids.eq(cls_token_id)
+    return ~token_mask
+
+
+def _get_batch_distribution(
+    model: PreTrainedModel, batch: Dict[str, Tensor], temperature: float, idf: bool, special_tokens_map: Dict[str, int]
+) -> Tensor:
+    """Calculate a discrete probability distribution for a batch of examples according to the methodology described
+    in `InfoLM`_.
+
+    Args:
+        model:
+            Initialized model from HuggingFace's `transformers package.
+        batch:
+            An input batch dictionary containing ``input_ids`` and ``attention_mask``.
+        temperature:
+            A temperature for calibrating language modelling. For more information, please reference `InfoLM`_ paper.
+        max_length:
+            A maximum length of input sequences. Sequences longer than `max_length` are to be trimmed.
+        idf:
+            An indication of whether normalization using inverse document frequencies should be used.
+
+    Return:
+        A discrete probability distribution.
+    """
+    seq_len = batch["input_ids"].shape[1]
+    prob_distribution_batch_list: List[Tensor] = []
+    token_mask = _get_token_mask(
+        batch["input_ids"],
+        special_tokens_map["pad_token_id"],
+        special_tokens_map["sep_token_id"],
+        special_tokens_map["cls_token_id"],
+    )
+    for mask_idx in range(seq_len):
+        input_ids = batch["input_ids"].clone()
+        input_ids[:, mask_idx] = special_tokens_map["mask_token_id"]
+        logits_distribution = model(input_ids, batch["attention_mask"]).logits
+        # [batch_size, seq_len, vocab_size] -> [batch_size, vocab_size]
+        logits_distribution = logits_distribution[:, mask_idx, :]
+        prob_distribution = F.softmax(logits_distribution / temperature, dim=-1)
+        if idf:
+            prob_distribution *= batch["input_ids_idf"][:, mask_idx].unsqueeze(1)
+        prob_distribution_batch_list.append(prob_distribution.unsqueeze(1).cpu())  # [batch_size, 1, vocab_size]
+        # Clean from memory
+        del input_ids, logits_distribution, prob_distribution
+
+    prob_distribution_batch = torch.cat(prob_distribution_batch_list, dim=1)  # [batch_size, seq_len, vocab_size]
+    prob_distribution_batch = torch.einsum("bsv, bs -> bsv", prob_distribution_batch, token_mask)
```

Comment:
there seems to be a device-mismatch here when running on GPU
```suggestion
    prob_distribution_batch = torch.einsum("bsv, bs -> bsv", prob_distribution_batch.to(token_mask.device), token_mask)
```